src/internal/storage/fileset/validator.go

```diff
@@ -0,0 +1,61 @@
+package fileset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
+)
+
+var _ FileSet = &pathValidator{}
+
+type pathValidator struct {
+	fs FileSet
+}
+
+// NewPathValidator validates the paths in a file set.
+func NewPathValidator(fs FileSet) FileSet {
+	return &pathValidator{
+		fs: fs,
+	}
+}
+
+func (pv *pathValidator) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
+	ctx = pctx.Child(ctx, "pathIterator")
+	var prev *index.Index
+	if err := pv.fs.Iterate(ctx, func(f File) error {
+		idx := f.Index()
+		if retErr := CheckIndex(prev, idx); retErr != nil {
```

Comment:
nit, `retErr` -> `err`.

Reply:
Is it ok to have same name `err` in different scopes? I know they do not conflict, just want to make sure if that's wanted.

Reply:
Yes, that is actually desirable. Taking advantage of scopes and using the conventional names for common variables (i.e. `err` and `ctx`) is a good practice. It reduces the probability of mistakes because it narrows the scope of consideration for these common variables.